### DISCUSSION TITLE
"content-addressable" models + relationship

### DISCUSSION THREAD

**Discussion by tamird** [sqlalchemy/sqlalchemy#11594] [Usage Questions]
Consider [this example in the `association_proxy` docs](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#simplifying-association-objects) and imagine that we didn't want to store duplicate keywords in the database, so we added a unique constraint:
```python
class Keyword(Base):
    __tablename__ = "keyword"
    id: Mapped[int] = mapped_column(primary_key=True)
    keyword: Mapped[str] = mapped_column("keyword", String(64), unique=True)

    def __init__(self, keyword: str):
        self.keyword = keyword

    def __repr__(self) -> str:
        return f"Keyword({self.keyword!r})"
```
Now we want to run the sample code from the example:
```python
user = User("log")
for kw in (Keyword("new_from_blammo"), Keyword("its_big")):
    user.keywords.append(kw)
print(user.keywords)
```
followed by the same snippet for another user:
```python
user = User("barrel")
for kw in (Keyword("new_from_blammo"), Keyword("its_big")):
    user.keywords.append(kw)
print(user.keywords)
```
This (probably) won't work because the uniqueness constraint won't allow it, right? Is it possible to make this work, such that only the association objects are created?

**Comment by zzzeek** [comment]
this is mentioned in that section:

> One caveat with our example above is that because Keyword objects are created for each dictionary set operation, the example fails to maintain uniqueness for the Keyword objects on their string name, which is a typical requirement for a tagging scenario such as this one. For this use case the recipe [UniqueObject](https://www.sqlalchemy.org/trac/wiki/UsageRecipes/UniqueObject), or a comparable creational strategy, is recommended, which will apply a “lookup first, then create” strategy to the constructor of the Keyword class, so that an already existing Keyword is returned if the given name is already present.

UniqueObject is an old recipe but the basic idea is still what we have for this right now

**Reply by tamird** [reply]
Neat, thanks. Do you think it'd be possible to update that recipe with type annotations? It's quite difficult to understand what's going on otherwise.

EDIT: in particular I'm talking about the [final recipe](https://github.com/sqlalchemy/sqlalchemy/wiki/UniqueObject#another-unique-object-recipe) in that wiki entry.

**Reply by tamird** [reply]
Also: how do I do this kind of thing using async? If I just put an event listener on my `session.sync_session` then I get
```
>           raise exc.MissingGreenlet(
                "greenlet_spawn has not been called; can't call await_only() "
                "here. Was IO attempted in an unexpected place?"
            )
E           sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

/Users/tamird/Library/Caches/pypoetry/virtualenvs/common-hf-Ms37h-py3.12/lib/python3.12/site-packages/sqlalchemy/util/_concurrency_py3k.py:123: MissingGreenlet
```
when I try to query the database for existing objects.

**Reply by tamird** [reply]
I eventually got this working, but ran into a number of undocumented cases. For one: issuing queries in `before_attach` on an async connection produces the greenlet error above.

I was able to work around it by doing more complicated work in `before_flush` instead, which did allow me to run queries (presumably because flushing runs on the greenlet when it is triggered by async queries on the session).

It would certainly be good to bolster the documentation [here](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#examples-of-event-listeners-with-async-engines-sessions-sessionmakers) to explain that e.g. `before_attach` can't issue queries (and in general make the events documentation exhaustive in terms of which events run on the greenlet and which ones don't).

**Reply by zzzeek** [reply]
sure, the recipe (the first part) is old and uses patterns more oriented towards a time when SQLAlchemy encouraged things like thread-local sessions.    There's ways to do a similar "implicit session" with async but overall I will always recommend more explicit patterns.

that second example at the bottom with the before_attach is more likely the better general idea here since it doesnt use any kind of thread/coroutine local stuff, but yes this would not be compatible with async because we dont consider `session.add()` to be awaitable.

For me personally I'd probably not rely upon association proxy creator here.  I'd not want my code performing a SELECT for every append operation, and I'd probably use a routine that pre-selects Keyword objects into a keyed collection of some kind and then matches them in bulk to the objects I'm inserting.     automating this into a before_flush could be feasible though I'd worry it's not easy to generalize to more complex cases, which is why this stuff is not any major documented section, it has to be tailored custom to the use case at hand.   The identity map architecture works against this pattern of on-the-fly objects that need to be uniqued.    A "once and for all" fix for this issue might rely upon a major new architectural feature that changes the behavior of the identity map for some classes of objects in some way.    a foolproof way to do this still eludes me however

**Comment by zzzeek** [comment]
looking at UniqueObject I'd be using the first recipe here.  Here is the first UniqueObject recipe using the UserKeyword mapping with asyncio:


```py
from __future__ import annotations

import asyncio
from typing import List
from typing import Optional

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.associationproxy import AssociationProxy
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import sessionmaker


Base = declarative_base()

engine = create_engine("sqlite://", echo=True)

Session = sessionmaker(bind=engine)


def _unique(session, cls, hashfunc, queryfunc, constructor, arg, kw):
    cache = getattr(session, "_unique_cache", None)
    if cache is None:
        session._unique_cache = cache = {}

    key = (cls, hashfunc(*arg, **kw))
    if key in cache:
        return cache[key]
    else:
        with session.no_autoflush:
            stmt = select(cls)
            stmt = queryfunc(stmt, *arg, **kw)
            obj = session.scalars(stmt).one_or_none()
            if not obj:
                obj = constructor(*arg, **kw)
                session.add(obj)
        cache[key] = obj
        return obj


class UniqueMixin:
    @classmethod
    def unique_hash(cls, *arg, **kw):
        raise NotImplementedError()

    @classmethod
    def unique_filter(cls, query, *arg, **kw):
        raise NotImplementedError()

    @classmethod
    async def async_as_unique(cls, async_session: AsyncSession, *arg, **kw):
        return await async_session.run_sync(cls.as_unique, *arg, **kw)

    @classmethod
    def as_unique(cls, session, *arg, **kw):
        return _unique(
            session, cls, cls.unique_hash, cls.unique_filter, cls, arg, kw
        )


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(64))

    user_keyword_associations: Mapped[List[UserKeywordAssociation]] = (
        relationship(
            back_populates="user",
            cascade="all, delete-orphan",
        )
    )

    # association proxy of "user_keyword_associations" collection
    # to "keyword" attribute
    keywords: AssociationProxy[List[Keyword]] = association_proxy(
        "user_keyword_associations",
        "keyword",
        creator=lambda keyword_obj: UserKeywordAssociation(
            keyword=keyword_obj
        ),
    )

    def __init__(self, name: str):
        self.name = name


class UserKeywordAssociation(Base):
    __tablename__ = "user_keyword"
    user_id: Mapped[int] = mapped_column(
        ForeignKey("user.id"), primary_key=True
    )
    keyword_id: Mapped[int] = mapped_column(
        ForeignKey("keyword.id"), primary_key=True
    )
    special_key: Mapped[Optional[str]] = mapped_column(String(50))

    user: Mapped[User] = relationship(
        back_populates="user_keyword_associations"
    )

    keyword: Mapped[Keyword] = relationship()


class Keyword(UniqueMixin, Base):
    __tablename__ = "keyword"
    id: Mapped[int] = mapped_column(primary_key=True)
    keyword: Mapped[str] = mapped_column("keyword", String(64))

    @classmethod
    def unique_hash(cls, keyword):
        return keyword

    @classmethod
    def unique_filter(cls, query, keyword):
        return query.where(Keyword.keyword == keyword)

    def __init__(self, keyword: str):
        self.keyword = keyword

    def __repr__(self) -> str:
        return f"Keyword({self.keyword!r})"


async def main():
    engine = create_async_engine("sqlite+aiosqlite://", echo=True)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as session:
        for username in "u1", "u2" "u3":
            user = User(username)
            for kw in (
                await Keyword.async_as_unique(
                    session, keyword="new_from_blammo"
                ),
                await Keyword.async_as_unique(session, keyword="its_big"),
            ):
                user.keywords.append(kw)
            print(user.keywords)

        await session.commit()


asyncio.run(main())

```

**Reply by tamird** [reply]
Does SQLAlchemy implement automatic bulk inserts? I.e. if the session has N objects of type T, will it insert them in a single query?

If yes, then this solution results in unbounded query amplification :(

**Reply by zzzeek** [reply]
the unit of work has a lot of logic that tries as hard as possible to group INSERTs into bulk operations, particularly in 2.0 where we have replaced "executemany" with "insertmanyvalues".  

I think you mean "no", re: unbounded query amplification

**Reply by tamird** [reply]
This solution produces an additional query per record, so I think I didn't mean "no".

**Reply by zzzeek** [reply]
i have no idea how if the unit of work needs to INSERT 5 records, that it would be preferable for it to emit five individual INSERT statements rather than just one.  can you clarify?

**Reply by tamird** [reply]
Sorry, I'm specifically talking about the unique record recipe above. For every unique record, it must query the database before instantiating the record. That's a query per record in addition to what the unit of work will later do on commit. It's unbounded because it turns one query into 1+ N where N is the number of new unique records.

**Reply by zzzeek** [reply]
I guess you mean that the batching, which would normally be available, is *lost*, if you dont pre-fetch these keywords, and autoflush is left turned on. 

Usually the model of "keywords" vs. records indicates millions of records for only a few hundred keywords, and we are already doing a SELECT for each keyword as well.

If I had hundreds of User records I would pre-fetch / create the Keyword records ahead of time in bulk.     

Overall, if I'm concerned about performance, I would be not using session events or any tricks like that and I'd be adding explicit logic to pre-query for what needs to be present in the most efficient way possible.

**Reply by zzzeek** [reply]
here's that, run it with echoing turned on to see the bulk take place.  The `_populate` function here can be further generalized to do the whole operation in one step rather than the querying ahead of time.

```py
from __future__ import annotations

import asyncio
from typing import List
from typing import Optional

from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.ext.associationproxy import association_proxy
from sqlalchemy.ext.associationproxy import AssociationProxy
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import sessionmaker


Base = declarative_base()

engine = create_engine("sqlite://", echo=True)

Session = sessionmaker(bind=engine)


def _populate(session, cls, hashfunc, objects):
    cache = session.info.get("_unique_cache", None)
    if cache is None:
        session.info['_unique_cache'] = cache = {}
    cache.update({(cls, hashfunc(obj)): obj for obj in objects})


def _unique(session, cls, hashfunc, queryfunc, constructor, arg, kw):
    cache = session.info.get("_unique_cache", None)
    if cache is None:
        session.info['_unique_cache'] = cache = {}

    key = (cls, hashfunc(*arg, **kw))
    if key in cache:
        return cache[key]
    else:
        with session.no_autoflush:
            stmt = select(cls)
            stmt = queryfunc(stmt, *arg, **kw)
            obj = session.scalars(stmt).one_or_none()
            if not obj:
                obj = constructor(*arg, **kw)
                session.add(obj)
        cache[key] = obj
        return obj


class UniqueMixin:
    @classmethod
    def unique_hash(cls, *arg, **kw):
        raise NotImplementedError()

    @classmethod
    def unique_filter(cls, query, *arg, **kw):
        raise NotImplementedError()

    @classmethod
    async def async_as_unique(cls, async_session: AsyncSession, *arg, **kw):
        return await async_session.run_sync(cls.as_unique, *arg, **kw)

    @classmethod
    def as_unique(cls, session, *arg, **kw):
        return _unique(
            session, cls, cls.unique_hash, cls.unique_filter, cls, arg, kw
        )


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "user"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(64))

    user_keyword_associations: Mapped[List[UserKeywordAssociation]] = (
        relationship(
            back_populates="user",
            cascade="all, delete-orphan",
        )
    )

    # association proxy of "user_keyword_associations" collection
    # to "keyword" attribute
    keywords: AssociationProxy[List[Keyword]] = association_proxy(
        "user_keyword_associations",
        "keyword",
        creator=lambda keyword_obj: UserKeywordAssociation(
            keyword=keyword_obj
        ),
    )

    def __init__(self, name: str):
        self.name = name


class UserKeywordAssociation(Base):
    __tablename__ = "user_keyword"
    user_id: Mapped[int] = mapped_column(
        ForeignKey("user.id"), primary_key=True
    )
    keyword_id: Mapped[int] = mapped_column(
        ForeignKey("keyword.id"), primary_key=True
    )
    special_key: Mapped[Optional[str]] = mapped_column(String(50))

    user: Mapped[User] = relationship(
        back_populates="user_keyword_associations"
    )

    keyword: Mapped[Keyword] = relationship()


class Keyword(UniqueMixin, Base):
    __tablename__ = "keyword"
    id: Mapped[int] = mapped_column(primary_key=True)
    keyword: Mapped[str] = mapped_column("keyword", String(64))

    @classmethod
    def unique_hash(cls, keyword):
        return keyword

    @classmethod
    def unique_filter(cls, query, keyword):
        return query.where(Keyword.keyword == keyword)

    def __init__(self, keyword: str):
        self.keyword = keyword

    def __repr__(self) -> str:
        return f"Keyword({self.keyword!r})"


async def main():
    engine = create_async_engine("sqlite+aiosqlite://", echo=True)
    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as session:

        keywords_needed = {"new_from_blammo", "its_big"}
        keywords_we_have = set(
            await session.scalars(
                select(Keyword).where(Keyword.keyword.in_(keywords_needed))
            )
        )
        keywords_to_add = {
            Keyword(keyword=k)
            for k in keywords_needed.difference(keywords_we_have)
        }
        session.add_all(keywords_to_add)

        _populate(
            session,
            Keyword,
            lambda k: k.keyword,
            keywords_to_add.union(keywords_we_have),
        )

        for username in "u1", "u2" "u3":
            user = User(username)
            for kw in (
                await Keyword.async_as_unique(
                    session, keyword="new_from_blammo"
                ),
                await Keyword.async_as_unique(session, keyword="its_big"),
            ):
                user.keywords.append(kw)
            print(user.keywords)

        await session.commit()


asyncio.run(main())

```

an explicit form like this is easier to see the intent of the code and is less risky if you have complex object graphs where rewriting collections is not sufficient

**Reply by tamird** [reply]
> I think the example of keywords and users is just not useful as a stand in for my use case.

that's why there is no built in feature here because the problem is too open ended

> Anyway what I was trying to point out above is that the documentation of event hooks vis-a-vis queries under async is insufficient at best and misleading at worst.

PRs welcome thanks

**Reply by zzzeek** [reply]
note, I am updating the cache lookup to be like this:

```
    cache = session.info.get("_unique_cache", None)
    if cache is None:
        session.info['_unique_cache'] = cache = {}
```

so that it's shared among Sessions that are created on the fly when working with AsyncSession
